src/MongoDB/ServerDescription.c

```diff
@@ -0,0 +1,234 @@
+
+	RETVAL_LONG((zend_long) mongoc_server_description_round_trip_time(intern->server_description));
+} /* }}} */
+
+/* {{{ proto integer MongoDB\Driver\ServerDescription::getType()
+   Returns the server’s node type */
+static PHP_METHOD(ServerDescription, getType)
+{
+	php_phongo_serverdescription_t* intern;
+
+	intern = Z_SERVERDESCRIPTION_OBJ_P(getThis());
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	RETVAL_LONG(php_phongo_server_description_type(intern->server_description));
+} /* }}} */
+
+/* {{{ MongoDB\Driver\ServerDescription function entries */
+ZEND_BEGIN_ARG_INFO_EX(ai_ServerDescription_void, 0, 0, 0)
+ZEND_END_ARG_INFO()
+
+static zend_function_entry php_phongo_serverdescription_me[] = {
+	/* clang-format off */
+	PHP_ME(ServerDescription, getHelloResponse, ai_ServerDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	PHP_ME(ServerDescription, getHost, ai_ServerDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	PHP_ME(ServerDescription, getLastUpdateTime, ai_ServerDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	PHP_ME(ServerDescription, getPort, ai_ServerDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	PHP_ME(ServerDescription, getRoundTripTime, ai_ServerDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	PHP_ME(ServerDescription, getType, ai_ServerDescription_void, ZEND_ACC_PUBLIC | ZEND_ACC_FINAL)
+	ZEND_NAMED_ME(__construct, PHP_FN(MongoDB_disabled___construct), ai_ServerDescription_void, ZEND_ACC_PRIVATE | ZEND_ACC_FINAL)
+	PHP_FE_END
+	/* clang-format on */
+};
+/* }}} */
+
+/* {{{ MongoDB\Driver\ServerDescription object handlers */
+static zend_object_handlers php_phongo_handler_serverdescription;
+
+static void php_phongo_serverdescription_free_object(zend_object* object) /* {{{ */
+{
+	php_phongo_serverdescription_t* intern = Z_OBJ_SERVERDESCRIPTION(object);
+
+	zend_object_std_dtor(&intern->std);
+
+	if (intern->properties) {
+		zend_hash_destroy(intern->properties);
+		FREE_HASHTABLE(intern->properties);
+	}
+
+	if (intern->server_description) {
+		mongoc_server_description_destroy(intern->server_description);
+	}
+}
+
+static zend_object* php_phongo_serverdescription_create_object(zend_class_entry* class_type) /* {{{ */
+{
+	php_phongo_serverdescription_t* intern = NULL;
+
+	intern = PHONGO_ALLOC_OBJECT_T(php_phongo_serverdescription_t, class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+
+	intern->std.handlers = &php_phongo_handler_serverdescription;
+
+	return &intern->std;
+} /* }}} */
+
+HashTable* php_phongo_serverdescription_get_properties_hash(phongo_compat_object_handler_type* object, bool is_debug) /* {{{ */
+{
+	php_phongo_serverdescription_t* intern = NULL;
+	zval                            retval = ZVAL_STATIC_INIT;
+
+	intern = Z_OBJ_SERVERDESCRIPTION(PHONGO_COMPAT_GET_OBJ(object));
+
+	php_phongo_server_description_to_zval(&retval, intern->server_description);
+
+	return Z_ARRVAL(retval);
+} /* }}} */
```

Comment:
I'm not completely sure I did this part correctly - I'm getting the expected outputs in the `debug` and `var_export` tests, but I'm no longer using the `is_debug` parameter in this function.

Reply:
Contrast your implementation here with `get_properties_hash` implementations for other classes. It looks like you modeled this function after `php_phongo_server_get_debug_info`. Server objects don't have `get_properties` handlers, which is why there is no direct equivalent of this function there.

Since the `get_debug_info` and `get_properties` handlers both expect to return `HashTable*`, you're not going to be able to use `php_phongo_server_description_to_zval` as written. That's wouldn't work with the `PHONGO_GET_PROPERTY_HASH_INIT_PROPS` macro, which I'll note is also the only thing that utilizes the `properties` field on `php_phongo_serverdescription_t` (not being used elsewhere in this PR AFAICT).

There are two different design models going on in PHPC:

1. For classes that implement both `get_debug_info` and `get_properties`, we use a common `get_properties_hash` function and a `PHONGO_GET_PROPERTY_HASH_INIT_PROPS` macro. This macro either allocates a new HashTable (for debug output) or gets a pointer to the internal HashTable on the `intern` struct (initializing it if needed). The logic there is that debug output (i.e. `get_debug_info`) always get a fresh HashTable, which is temporary and can be freed by PHP (also why we set `*is_temp = 1;` in various `get_debug_info` handlers). For `get_properties`, the HashTable doesn't need to be temporary (we can retain ownership of it) so we just return our own internal HashTable.
2. For classes that skip `get_properties`, everything happens in `get_debug_info` and there's no need for a shared utility function. In the case of Server, we _could_ allocate a HashTable to populate and return but since we already needed a `server_to_zval` function to call from Manager's own debug handler, we instead declare a local zval and use the same function. At the end of `php_phongo_server_get_debug_info` we use `Z_ARRVAL` to just return the HashTable within the zval since the top-level zval itself isn't relevant. We're not leaking memory by doing so since the dynamic memory allocation all pertains to the HashTable anyway (zval is just a local struct otherwise).

Since it'd be nice to have both handlers for ServerDescription, I think it makes sense to alter `php_phongo_server_description_to_zval` to instead populate a HashTable directly instead of taking a zval and operating on _its_ HashTable. That will mean TopologyDescription's debug handler will differ a bit from Manager's (since they're analogous to one another in the sends that they both contain many server descriptions).

Let me know if you have any questions above the above or would like me to elaborate further on something.

Reply:
Got it! I've updated the code according to the first design model you described already, except that I populated the hashtable directly in `get_properties_hash` and removed the call to `php_phongo_server_description_to_zval`. Should I refactor this to populate the hashtable in `php_phongo_server_description_to_zval` then?

Reply:
Populating it directly in `get_properties_hash` makes sense for this PR. If you end up needing to re-use the code for TopologyDescription you can factor it out in that PR.